README.md
# runibic: UniBic biclustering algorithm for R

This package contains implementation of UniBic biclustering algorithm for gene expression data [Wang2016]
The algorithm tries to locate trend-preserving biclusters within complex and noisy data.

## Functions
This package provides the following main functions:
* `BCUnibic`/`runibic` - parallel UniBic for continuous data
* `BCUnibicD` - parallel UniBic for discrete data

The package provides some additional functions:
* `pairwiseLCS` - calculates Longest Common Subsequence (LCS) between two vectors
* `calculateLCS` - calculates LCSes between all pairs of the input dataset
* `backtrackLCS` - recovers LCS from the dynamic programming matrix
* `cluster` - main part of UniBic algorithm (biclusters seeding and expanding)
* `unisort` - returns matrix of indexes based on the increasing order in each row
* `discretize` - performs discretization using Fibonacci heap (sorting method used originally in UniBic) or standard sorting


## Installation
The package may be installed as follows:
```r
install.packages("devtools")
devtools::install_github("athril/runibic")
```

## Example
### Gene expression dataset
This example presents how to use runibic package on gene expression dataset:
```r
library(runibic)
library(biclust)
data(BicatYeast)
res <- biclust(method=BCUnibic(),BicatYeast)
drawHeatmap(BicatYeast, res, 1)
parallelCoordinates(BicatYeast,res,1)
```

### Summarized experiment
This example presents how to use runibic package on SummarizedExperiment:
```r
library(runibic)
library(biclust)
library(SummarizedExperiment)
data(airway, package="airway")
se <- airway[1:20,]
res<- runibic(se)
parallelCoordinates(assays(se)[[1]], res[[1]], 2)
```

## Tutorial
Please check [runibic tutorial](https://github.com/athril/runibic/tree/master/vignettes/runibic.Rmd)

## Citation
For the original sequential version of the UniBic please use the following citation:

**Zhenjia Wang, Guojun Li, Robert W. Robinson, Xiuzhen Huang
*UniBic: Sequential row-based biclustering algorithm for analysis of gene expression data*
Scientific Reports 6, 2016; 23466, doi: https://doi:10.1038/srep23466**


If you use in your work this package with parallel version of UniBic please use the following citation:

**Patryk Orzechowski, Artur Pańszczyk, Xiuzhen Huang Jason H. Moore:
*runibic: a Bioconductor package for parallel row-based biclustering of gene expression data*
bioRxiv, 2017; 210682, doi: [https://doi.org/10.1101/210682](https://doi.org/10.1101/210682)**

BibTex entry:
```
@article{orzechowski2018runibic,
  author = {Orzechowski, Patryk and Pańszczyk, Artur and Huang, Xiuzhen and Moore, Jason H},
  title = {runibic: a Bioconductor package for parallel row-based biclustering of gene expression data},
  journal = {Bioinformatics},
  volume = {},
  number = {},
  pages = {bty512},
  year = {2018},
  doi = {10.1093/bioinformatics/bty512},
  URL = {http://dx.doi.org/10.1093/bioinformatics/bty512},
  eprint = {/oup/backfile/content_public/journal/bioinformatics/pap/10.1093_bioinformatics_bty512/4/bty512.pdf}
}
```


## References
* [Wang2016] Wang, Zhenjia, et al. "UniBic: Sequential row-based biclustering algorithm for analysis of gene expression data." Scientific reports 6 (2016): 23466.


NEWS
Version 1.2.5
Removing warnings.

Version 1.2.4
Updating references.

Version 1.2.3
Improvement of results for sets with a small number of columns.

Version 1.2.2
Fixing build warning/errors on Windows/OSX.

Version 1.2.1
Synchronizing the last change with Bioconductor 3.7

Version 1.0.3
Applying previous releases in the dev branch.

Version 1.0.2
Fix: modification of default parameters for small datasets in order to keep consistency between different sizes of the input.

Version 1.0.1
Fixing memory management bug.

Version 0.99.14
Updating tutorial, adding citation.

Version 0.99.13
Documentation to the package was updated. 

Version 0.99.9
Including required biocViews. Refactoring R code.

Version 0.99.8
Code optimizations.

Version 0.99.6
Parallel version using OpenMP.

Version 0.99.5
Fixing discretize() function bug.

Version 0.99.2
Improved performance of functions in package.
Removing some errors and warnings.

Version 0.99.1
Compatibility with SummarizedExperiment.
Providing documentation and examples.

Version 0.99.0
Development of the runibic package prototype completed.
First upload to Bioconductor.

Version 0.90.1
Working version of UniBic algorithm.






R/runibic.R
#' runibic: parallel row-based biclustering algorithm
#' for analysis of gene expression data in R
#' @docType package
#' @author Patryk Orzechowski \url{patryk.orzechowski@gmail.com}, Artur Pańszczyk \url{panszczyk.artur@gmail.com}
#' @name runibic
#' @useDynLib runibic
#' @importFrom Rcpp evalCpp
#' @import testthat SummarizedExperiment
#' @importFrom biclust biclust bicluster
#' @export runiDiscretize
#' @export set_runibic_params
#' @export runibic
#' @export BCUnibic
#' @export BCUnibicD
#' @description \code{\link{runibic}} is a package that contains much faster parallel version of one of the most accurate biclustering algorithms, UniBic.
#' The original method was reimplemented from C to C++11, OpenMP was added for parallelization.
#'
#' If you use this package, please cite it as:
#' Patryk Orzechowski, Artur Pańszczyk, Xiuzhen Huang, Jason H Moore; 
#' "runibic: a Bioconductor package for parallel row-based biclustering of gene expression data";
#' Bioinformatics, 2018, bty512,
#' doi: https://doi.org/10.1093/bioinformatics/bty512
#'
#' @details
#' For a given input matrix we first perform discretization and create index matrix using \code{\link{runiDiscretize}} function.
#' The discretization is performed taking into account quantiles of the data.
#' The resulting index matrix allows to detect order-preserving trends between each pair of the rows
#' irrespective to the order of columns.
#' After the ranking, the matrix is split by rows into subgroups based on the significance of the future biclusters.
#' In each of the chunks, we calculate pairwise calculations of Longest Common Subsequence LCS between all pairs of the rows.
#' LCS calculations are performed using dynamic programming and determine the longest order-preserving trend between each pair of the rows.
#' After partitioning the matrix strict order-preserving biclusters are determined and later expanded
#' to approximate-trend biclusters within \code{\link{cluster}} function.
#'
#' This package provides 3 main functions:
#' \code{\link{runibic}} and \code{\link{BCUnibic}} perform UniBic biclustering algorithm on numeric data, whilst
#' \code{\link{BCUnibicD}} could be applied to integer ones. The latter two methods are compatible with \code{\link[biclust]{Biclust}} class.
#'
#' @references Wang, Zhenjia, et al. "UniBic: Sequential row-based biclustering algorithm for analysis of gene expression data." Scientific reports 6 (2016): 23466.
#' @references Patryk Orzechowski, Artur Pańszczyk, Xiuzhen Huang, Jason H. Moore: "runibic: a Bioconductor package for parallel row-based biclustering of gene expression data", bioRxiv (2017): 210682, doi: https://doi.org/10.1101/210682
NULL


#' Class BCUnibic
#'
#' An S4 class to represent \code{\link{BCUnibic-class}} UniBic biclustering algorithm
#' for numeric input. The class is intended to use with 
#'
#' @name BCUnibic-class
#' @rdname BCUnibic-class
#' @seealso \code{\link{runibic}} 
setClass(Class = "BCUnibic", contains = "BiclustMethod", 
    prototype = prototype(biclustFunction = function(x, ...) {    
    runibic(x, ...)
}))


#' Class BCUnibicD
#'
#' An S4 class \code{\link{BCUnibicD-class}} defines UniBic biclustering algorithm
#' for discrete input.
#'
#' @name BCUnibicD-class
#' @rdname BCUnibicD-class
setClass(Class = "BCUnibicD", contains = "BiclustMethod",
    prototype = prototype(biclustFunction = function(x, ...) {
    runibic_d(x, ...)
}))


#' @describeIn runibic \code{\link{BCUnibic}} performs biclustering using UniBic on numeric matrix.
#' It is intended to use as a method called from \code{\link[biclust]{biclust}}.
BCUnibic <- function(x = NULL, t = 0.95, q = 0, f = 1, nbic = 100, div = 0, useLegacy = FALSE) {
    if (is.null(x))
        return(methods::new("BCUnibic"))
    res <- biclust(x = x, method = BCUnibic(), t = 0.95, q = 0, f = 1, nbic = 100, div = 0, useLegacy = FALSE)
    res@Parameters$Call <- match.call()
    return (res);
}


#' @describeIn runibic perform biclustering using UniBic on integer matrix.
#' It is intended to use as a method called from \code{\link[biclust]{biclust}}.
BCUnibicD <- function(x = NULL, t = 0.95, q = 0, f = 1, nbic = 100, div = 0, useLegacy = FALSE) {
    if (is.null(x))
        return(methods::new("BCUnibicD"))
    res <- biclust(x = x, method = BCUnibicD(), t = 0.95, q = 0, f = 1, nbic = 100, div = 0, useLegacy = FALSE)
    res@Parameters$Call <- match.call()
    return (res);
}


runibic_d <- function(x = NULL, t = 0.95, q = 0, f = 1, nbic = 100, div = 0, useLegacy = FALSE) {
    MYCALL <- match.call()
    
    set_runibic_params(t, q, f, nbic, div, useLegacy)
    iX <- unisort(x)
    LCSRes <- calculateLCS(x, TRUE)
    res <- cluster(iX, x, LCSRes$lcslen, LCSRes$a, LCSRes$b, nrow(x), ncol(x) )
    return(biclust::BiclustResult(as.list(MYCALL), matrix(unlist(res["RowxNumber"]), ncol = as.numeric(res["Number"]), byrow = FALSE),
        matrix(unlist(res["NumberxCol"]), nrow = as.numeric(res["Number"]), byrow = FALSE), as.numeric(res["Number"]),
        res["info"]))
}
#' runibic
#'
#' Each of the following functions \code{\link{BCUnibic}}, \code{\link{BCUnibicD}}, 
#' \code{\link{runibic}} perform biclustering
#' using UniBic biclustering algorithm. The major difference
#' between the functions is that \code{\link{BCUnibicD}} require a discretized matrix, 
#' whilst \code{\link{BCUnibic}} (or \code{\link{runibic}})
#' could be applied to numeric one.
#'
#' @param x numeric or integer matrix (depends on the function)
#' @param t consistency level of the block (0.5-1.0].
#' @param q a double value for quantile discretization
#' @param f filtering overlapping blocks (default 1 do not remove any blocks)
#' @param nbic maximum number of biclusters in output
#' @param div number of ranks for up(down)-regulated genes: default: 0==ncol(x)
#' @param useLegacy boolean value for using legacy parameter settings
#' @return \code{\link[biclust]{Biclust}} object with detected biclusters
#'
#' @usage runibic(x = NULL, t = 0.95, q = 0, f = 1, nbic = 100, div = 0, useLegacy=FALSE)
#' @seealso \code{\link{runiDiscretize}} \code{\link{set_runibic_params}} \code{\link{BCUnibic-class}} \code{\link{BCUnibicD-class}} \code{\link{unisort}}
#' @describeIn runibic perform biclustering using UniBic on numeric matrix.
#'
#' @examples
#' A <- matrix(replicate(100, rnorm(100)), nrow=100, byrow=TRUE)
#' runibic(A)
#' BCUnibic(A)
#' BCUnibic(A, t = 0.95, q = 0, f = 1, nbic = 100, div = 0)
#' B <- runiDiscretize(A)
#' runibic(B)
#' BCUnibicD(B, t = 0.95, q = 0, f = 1, nbic = 100, div = 0)
#' biclust::biclust(A, method=BCUnibic(), t = 0.95, q = 0, f = 1, nbic = 100, div = 0)
#' biclust::biclust(B, method=BCUnibicD(), t = 0.95, q = 0, f = 1, nbic = 100, div = 0)
runibic <- function(x = NULL, t = 0.95, q = 0, f = 1, nbic = 100, div = 0, useLegacy = FALSE) {
    if(inherits(x,"SummarizedExperiment")){
        x_d <- lapply(assays(x), runiDiscretize)
        return (lapply(x_d, runibic_d, t, q, f, nbic, div))
    }
    set_runibic_params(t, q, f, nbic, div, useLegacy)
    x_d <- runiDiscretize(x)
    return(runibic_d(x_d, t, q, f, nbic, div, useLegacy))
}


